balancer/grpclb/grpclb_util.go

```diff
@@ -60,32 +61,41 @@ import (
 // so when grpclb client lose contact with remote balancers, the parent
 // ClientConn's resolver will re-resolve.
 type lbManualResolver struct {
-	scheme string
-	ccr    resolver.ClientConn
-
-	ccb balancer.ClientConn
-}
-
-func (r *lbManualResolver) Build(_ resolver.Target, cc resolver.ClientConn, _ resolver.BuildOptions) (resolver.Resolver, error) {
-	r.ccr = cc
-	return r, nil
-}
-
-func (r *lbManualResolver) Scheme() string {
-	return r.scheme
-}
+	// These fields are setup at creation time and are read-only after that, and
+	// therefore need not be protected with a mutex.
+	resolver.Builder                      // The underlying manual resolver builder.
+	resolver.Resolver                     // The underlying manual resolver.
+	ccb               balancer.ClientConn // ClientConn passed to the parent channel.
+
+	// The resolver.ClientConn is updated everytime the resolver on the channel
+	// is restarted, and this happens when the channel exits IDLE.
+	mu  sync.Mutex
+	ccr resolver.ClientConn
+}
+
+func newManualResolver(scheme string, ccb balancer.ClientConn) *lbManualResolver {
+	mr := manual.NewBuilderWithScheme(scheme)
+	r := &lbManualResolver{
+		Builder:  mr,
+		Resolver: mr,
+		ccb:      ccb,
+	}
 
-// ResolveNow calls resolveNow on the parent ClientConn.
-func (r *lbManualResolver) ResolveNow(o resolver.ResolveNowOptions) {
-	r.ccb.ResolveNow(o)
+	mr.BuildCallback = func(_ resolver.Target, ccr resolver.ClientConn, _ resolver.BuildOptions) {
+		r.mu.Lock()
+		r.ccr = ccr
```

Comment:
Isn't this functionality (and `UpdateState`) handled by the manual resolver internals?  Or is there something different here that I'm missing?

Reply:
Yup, it is. I didn't realize it was being handled there. So, we don't this type anymore.